agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -1221,31 +1222,60 @@ func (dg *dockerGoClient) APIVersion() (dockerclient.DockerVersion, error) {
 	return dg.sdkClientFactory.FindClientAPIVersion(client), nil
 }
 
-// Stats returns a channel of *docker.Stats entries for the container.
-func (dg *dockerGoClient) Stats(id string, ctx context.Context) (<-chan *docker.Stats, error) {
-	client, err := dg.dockerClient()
+// Stats returns a channel of *types.Stats entries for the container.
+func (dg *dockerGoClient) Stats(id string, ctx context.Context) (<-chan *types.Stats, error) {
+	client, err := dg.sdkDockerClient()
 	if err != nil {
 		return nil, err
 	}
 
-	stats := make(chan *docker.Stats)
-	options := docker.StatsOptions{
-		ID:                id,
-		Stats:             stats,
-		Stream:            true,
-		Context:           ctx,
-		InactivityTimeout: StatsInactivityTimeout,
-	}
+	// Create channel to hold the stats
+	statsChnl := make(chan *types.Stats)
+	var resp types.ContainerStats
 
 	go func() {
-		statsErr := client.Stats(options)
-		if statsErr != nil {
-			seelog.Infof("DockerGoClient: Unable to retrieve stats for container %s: %v",
-				id, statsErr)
+		// ContainerStats outputs a io.ReadCloser and an OSType
+		stream := true
+		resp, err = client.ContainerStats(ctx, id, stream)
+		if err != nil {
+			seelog.Warnf("DockerGoClient: Unable to retrieve stats for container %s: %v", id, err)
+			close(statsChnl)
+			return
 		}
+
+		// handle inactivity timeout
+		_, cancelRequest := context.WithCancel(ctx)
```

Comment:
i think we need to pass the ctx returned here to client.ContainerStats(...) so the request can be cancelled when cancelRequest is invoked?

Reply:
That's changing the behavior compared to before when go-dockerclient used to handle this timeout. I'll stick to what I think is the as-is translation for now, I don't want the sdk migration to contain any behavior changes.

Reply:
i think go-dockerclient has this behavior, just under different level. In go-dockerclient, the inactivity timeout is handled in the [stream](https://github.com/aws/amazon-ecs-agent/blob/e8b5229a9ea8b748ebad4db4c098ccc8df2dcd5f/agent/vendor/github.com/fsouza/go-dockerclient/client.go#L521) function, in which it passes a subCtx when it actually calls docker engine via HTTPClient.Do [here](https://github.com/aws/amazon-ecs-agent/blob/e8b5229a9ea8b748ebad4db4c098ccc8df2dcd5f/agent/vendor/github.com/fsouza/go-dockerclient/client.go#L596), and the cancel function corresponds to the subCtx is called when the inactivity timeout happens so that the request will be canceled. In your implementation the inactivity timeout is handled in Stats function, and the request to docker is made via the docker api client.ContainerStats. In that case the client.ContainerStats call should be canceled when inactivity timeout happens, similar to what go-dockerclient does to HTTPClient.Do. 

Let me know if this doesn't make sense..